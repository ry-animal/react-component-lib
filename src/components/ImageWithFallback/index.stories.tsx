import { ComponentMeta, ComponentStory } from "@storybook/react";
import ImageWithFallback from ".";
import BrokenImage from "../../assets/images/broken-image.jpg";

export default {
  component: ImageWithFallback,
  title: "Components/ImageWithFallback",
} as ComponentMeta<typeof ImageWithFallback>;

const Template: ComponentStory<typeof ImageWithFallback> = (args) => (
  <ImageWithFallback style={{ maxWidth: "400px" }} {...args} />
);

const args = {
  imgSrc: "//source.unsplash.com/user/erondu/400x400",
  fallbackImgSrc: BrokenImage,
  alt: "img",
};

export const DefaultImage = Template.bind({});
DefaultImage.args = {
  ...args,
};

export const Fallback = Template.bind({});
Fallback.args = {
  ...args,
  imgSrc: "./bad-path.jpg",
};

export const Placeholder = Template.bind({});
Placeholder.args = {
  ...args,
  imgSrc: "//source.unsplash.com/user/erondu/4000x4000",
  placeholderImgSrc: "//source.unsplash.com/user/erondu/5x5",
};
