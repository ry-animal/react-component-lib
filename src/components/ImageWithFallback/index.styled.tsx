import styled from "styled-components";

export const Img = styled.img<{ placeholderImg?: string }>`
  ${({ placeholderImg }) =>
    placeholderImg &&
    `
    background-image: url("${placeholderImg}");
    background-size: cover;
  `}
`;
