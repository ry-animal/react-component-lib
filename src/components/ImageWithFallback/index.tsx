import { ImgHTMLAttributes, useEffect, useState } from "react";
import BrokenImage from "../../assets/images/broken-image.jpg";
import * as Styled from "./index.styled";

export interface ImageWithFallbackProps
  extends ImgHTMLAttributes<HTMLImageElement> {
  /**
   * Url src for img
   */
  imgSrc: string;
  /**
   * Img to show if default errors
   */
  fallbackImgSrc?: string;
  /**
   * Img to show if error
   */
  errorImgSrc?: string;
  /**
   * Img to show when loading
   */
  placeholderImgSrc?: string;
  /**
   *
   * Image Fit
   */
  imgFit?: "contain" | "cover";
}

export const ImageWithFallback = ({
  imgSrc,
  fallbackImgSrc = BrokenImage,
  errorImgSrc = BrokenImage,
  placeholderImgSrc,
  ...props
}: ImageWithFallbackProps) => {
  const [src, setSrc] = useState(imgSrc);
  const [loading, setLoading] = useState(true);
  const onError = () =>
    setSrc(src === fallbackImgSrc ? errorImgSrc : fallbackImgSrc);
  const onLoad = () => setLoading(false);

  useEffect(() => {
    setSrc(imgSrc);
    setLoading(true);
  }, [imgSrc]);

  return (
    <Styled.Img
      src={src}
      onError={onError}
      onLoad={onLoad}
      placeholderImg={loading ? placeholderImgSrc : undefined}
      {...props}
    />
  );
};

export default ImageWithFallback;
