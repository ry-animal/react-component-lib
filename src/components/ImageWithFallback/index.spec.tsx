import { fireEvent, render, screen } from "@testing-library/react";
import "jest-styled-components";
import ImageWithFallback from ".";

describe("<ImageWithFallback />", () => {
  const sampleProps = () => ({
    imgSrc: "//source.unsplash.com/user/erondu/800x800",
    fallbackImgSrc: "./fallback-img.jpg",
    errorImgSrc: "./error-loading-images.jpg",
    alt: "img",
  });
  describe("rendering", () => {
    test("Default rendering matches snapshot", () => {
      expect(
        render(<ImageWithFallback {...sampleProps()} />).asFragment()
      ).toMatchSnapshot();
    });

    test("Renders fallback image when src broken", () => {
      expect(
        render(
          <ImageWithFallback {...sampleProps()} imgSrc={"./bad-path.jpg"} />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("Renders placeholder when loading if provided", () => {
      expect(
        render(
          <ImageWithFallback
            {...sampleProps()}
            placeholderImgSrc="//source.unsplash.com/user/erondu/5x5"
          />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("Should render fallback img if bad img src", async () => {
      const props = sampleProps();
      render(
        <ImageWithFallback
          {...props}
          placeholderImgSrc="//source.unsplash.com/user/erondu/5x5"
        />
      );
      let img = screen.getByAltText(`${props.alt}`);
      expect(img).toHaveAttribute("src", props.imgSrc);

      fireEvent.error(img);

      img = screen.getByAltText(`${props.alt}`);
      expect(img).toHaveAttribute("src", props.fallbackImgSrc);
    });

    test("Show placeholder when loading, then remove when loaded", async () => {
      const props = sampleProps();
      const placeholderSrc = "//source.unsplash.com/user/erondu/5x5";
      const backgroundImgStyle = `url("${placeholderSrc}")`;
      render(
        <ImageWithFallback {...props} placeholderImgSrc={placeholderSrc} />
      );
      let img = screen.getByAltText(`${props.alt}`);

      expect(img).toHaveStyleRule("background-image", backgroundImgStyle);

      fireEvent.load(img);

      img = screen.getByAltText(`${props.alt}`);
      expect(img).not.toHaveStyleRule("background-image", backgroundImgStyle);
    });

    test("Should render errorImgSrc if fallback img also bad", async () => {
      const props = sampleProps();
      render(
        <ImageWithFallback
          {...props}
          placeholderImgSrc="//source.unsplash.com/user/erondu/5x5"
        />
      );
      let img = screen.getByAltText(`${props.alt}`);
      expect(img).toHaveAttribute("src", props.imgSrc);

      fireEvent.error(img);

      img = screen.getByAltText(`${props.alt}`);
      expect(img).toHaveAttribute("src", props.fallbackImgSrc);

      fireEvent.error(img);

      img = screen.getByAltText(`${props.alt}`);
      expect(img).toHaveAttribute("src", props.errorImgSrc);
    });
  });
});
