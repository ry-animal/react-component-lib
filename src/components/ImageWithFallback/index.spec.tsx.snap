// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageWithFallback /> rendering Default rendering matches snapshot 1`] = `
<DocumentFragment>
  <img
    alt="img"
    class=""
    src="//source.unsplash.com/user/erondu/800x800"
  />
</DocumentFragment>
`;

exports[`<ImageWithFallback /> rendering Renders fallback image when src broken 1`] = `
<DocumentFragment>
  <img
    alt="img"
    class=""
    src="./bad-path.jpg"
  />
</DocumentFragment>
`;

exports[`<ImageWithFallback /> rendering Renders placeholder when loading if provided 1`] = `
<DocumentFragment>
  .c0 {
  background-image: url("//source.unsplash.com/user/erondu/5x5");
  background-size: cover;
}

<img
    alt="img"
    class="c0"
    src="//source.unsplash.com/user/erondu/800x800"
  />
</DocumentFragment>
`;
